src/transactions/RestoreFootprintOpFrame.cpp

```diff
@@ -0,0 +1,151 @@
+{
+    return header.ledgerVersion >= 20;
+}
+
+bool
+RestoreFootprintOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    throw std::runtime_error("RestoreFootprintOpFrame::doApply needs Config");
+}
+
+bool
+RestoreFootprintOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
+                                 Hash const& sorobanBasePrngSeed)
+{
+    RestoreFootprintMetrics metrics(app.getMetrics());
+
+    auto const& resources = mParentTx.sorobanResources();
+    auto const& footprint = resources.footprint;
+
+    UnorderedMap<LedgerKey, uint32_t> originalExpirations;
+
+    for (auto const& lk : footprint.readWrite)
+    {
+        auto ltxe = ltx.load(lk);
+        if (ltxe)
+        {
+            auto keySize = xdr::xdr_size(lk);
+            auto entrySize = xdr::xdr_size(ltxe.current());
+            metrics.mLedgerWriteByte += keySize;
+            metrics.mLedgerWriteByte += entrySize;
+
+            // Divide the limit by 2 for the metadata check since the previous
+            // state is also included in the meta.
+            if (resources.extendedMetaDataSizeBytes / 2 <
+                    metrics.mLedgerWriteByte ||
+                resources.writeBytes < metrics.mLedgerWriteByte ||
+                resources.readBytes < metrics.mLedgerWriteByte)
+            {
+                innerResult().code(RESTORE_FOOTPRINT_RESOURCE_LIMIT_EXCEEDED);
+                return false;
+            }
+
+            originalExpirations.emplace(LedgerEntryKey(ltxe.current()),
+                                        getExpirationLedger(ltxe.current()));
+
+            auto ledgerSeq = ltx.loadHeader().current().ledgerSeq;
+            auto const& expirationSettings = app.getLedgerManager()
+                                                 .getSorobanNetworkConfig(ltx)
+                                                 .stateExpirationSettings();
+            auto minPersistentExpirationLedger =
+                ledgerSeq + expirationSettings.minRestorableEntryExpiration;
+
+            if (getExpirationLedger(ltxe.current()) <
+                minPersistentExpirationLedger)
+            {
+                setExpirationLedger(ltxe.current(),
+                                    minPersistentExpirationLedger);
+            }
+        }
+    }
+
+    mParentTx.pushInitialExpirations(std::move(originalExpirations));
+
+    innerResult().code(RESTORE_FOOTPRINT_SUCCESS);
+    return true;
+}
+
+bool
+RestoreFootprintOpFrame::doCheckValid(SorobanNetworkConfig const& config,
+                                      uint32_t ledgerVersion)
+{
+    auto const& footprint = mParentTx.sorobanResources().footprint;
+    if (!footprint.readOnly.empty())
+    {
+        innerResult().code(RESTORE_FOOTPRINT_MALFORMED);
+        return false;
+    }
+
+    for (auto const& lk : footprint.readWrite)
+    {
+        if (!isSorobanEntry(lk) || isTemporaryEntry(lk))
+        {
+            innerResult().code(RESTORE_FOOTPRINT_MALFORMED);
```

Comment:
Should be `!isSorobanDataEntry` since we should only support `DATA_ENTRY` keys.

Reply:
Good catch. Fixed that same bug in the bump op as well